### PR TITLE
ci(gcb): add --docker flag

### PR DESCRIPTION
**Pull request by devjgm**
This PR adds the `-d | --docker` flags to run the specified build using
the local docker. This let's users run our GCB builds locally in an
environment that's very similar (exact? pretty close) to the GCB
environment. This also makes debugging easier. Users can start a bash
shell in the specified docker container by passing the `-s | --docker_shell`
flag.

Examples:

```
build.sh cmake-install  # Runs build on GCB
build.sh cmake-install --local  # Runs build on your local machine
build.sh cmake-install --docker  # Runs build in docker
build.sh cmake-install --docker-shell  # Starts shell in docker
```

These docker builds create a "build-out" directory on the host system to
store all their output (similar to our current "cmake-out" directory,
but not cmake-specific). All their build output is stored in a
build-specific dir w/in build-out/.

Coming soon... gcloud ADC credentials will be available in the container so that integration tests work.

---
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/6122)

### CODE REVIEW

File: ci/cloudbuild/build.sh [hunk 1]
```diff
@@ -83,14 +97,25 @@ while true; do
...
 fi
 readonly BUILD_NAME="$1"
 
+# --local is the most fundamental build mode, in that all other builds
+# eventually call this one. For example, a --docker build will build the
+# specified docker image, then in a container from that image it will run the
+# --local build. Similarly, the GCB build will submit the build to GCB, which
+# will call the --local build.
 if [[ "${LOCAL_BUILD}" = "true" ]]; then
+  test -n "${DISTRO}" && io::log_red "Ignoring --distro for local build"
```
Comment (devbww) [inline ci/cloudbuild/build.sh:113]: Perhaps s/--distro/--distro=${DISTRO}/ for more information.

File: ci/cloudbuild/build.sh [hunk 2]
```diff
@@ -109,6 +134,44 @@ if [[ -z "${DISTRO}" ]]; then
...
+    # Mounts an empty volume over "build-out" to isolate builds from each
+    # other. Doesn't affect GCB builds, but it helps our local docker builds.
+    "--volume=/workspace/build-out"
+    "--volume=${PROJECT_ROOT}:/workspace:Z"
+    "--workdir=/workspace"
+    "--volume=${out_cmake}:/workspace/cmake-out:Z"
+    "--volume=${out_home}:/h:Z"
+    "--env=HOME=/h"
+  )
+  cmd=(./ci/cloudbuild/build.sh --local "${BUILD_NAME}")
```
Comment (devbww) [inline ci/cloudbuild/build.sh:165]: s:./::

**Review by devbww — APPROVED** [review]

File: ci/cloudbuild/build.sh (hunk 1 above)
Comment (devjgm) [reply ci/cloudbuild/build.sh]: Done. Though reworded to keep it on one line ;-)

File: ci/cloudbuild/build.sh (hunk 2 above)
Comment (devjgm) [reply ci/cloudbuild/build.sh]: Done.

**Pull request merged**